File: packages/basic_message_channel_example/ios/Classes/BasicMessageChannelExamplePlugin.swift
import Flutter
import UIKit
import AVFoundation

public class BasicMessageChannelExamplePlugin: NSObject, FlutterPlugin {
    private let channel: FlutterBasicMessageChannel
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        // Setup basic message channel to communicate with flutter part
        let channel = FlutterBasicMessageChannel(
            name: "basic_message_channel_example",
            binaryMessenger: registrar.messenger(),
            codec: FlutterStringCodec.sharedInstance()
        )
        let instance = BasicMessageChannelExamplePlugin(channel: channel)
        instance.setupChannel()
        instance.startOrientationObserver()
        registrar.publish(instance)
    }
    
    init(channel: FlutterBasicMessageChannel) {
        self.channel = channel
    }
    
    // Setup a handler for messages from the flutter part
    private func setupChannel() {
        channel.setMessageHandler {  [weak self] message, reply in
            if message as? String == "get_orientation" {
                reply(self?.currentOrientation())
            } else {
                reply("Unknown command")
            }
        }
    }

    // Get current orientation
    private func currentOrientation() -> String {
        let orientation = UIApplication.shared.windows.first?.windowScene?.interfaceOrientation
        switch orientation {
        case .portrait, .portraitUpsideDown:
            return "Portrait"
        case .landscapeLeft, .landscapeRight:
            return "Landscape"
        default:
            return "Undefined"
        }
    }

    // Listen for orientation change
    private func startOrientationObserver() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(orientationChanged),
            name: UIDevice.orientationDidChangeNotification,
            object: nil
        )
        UIDevice.current.beginGeneratingDeviceOrientationNotifications()
    }

    // Receive orientation change notification
    @objc private func orientationChanged() {
        let info = currentOrientation()
        // Send message to flutter part
        channel.sendMessage(String(format: "orientation_update:%@", info))
    }
}

File: packages/basic_message_channel_example/android/src/main/kotlin/com/example/basic_message_channel_example/BasicMessageChannelExamplePlugin.kt
package com.example.basic_message_channel_example

import android.content.Context
import android.content.res.Configuration
import android.view.OrientationEventListener
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BasicMessageChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.StringCodec

/** BasicMessageChannelExamplePlugin */
class BasicMessageChannelExamplePlugin : FlutterPlugin {
    private lateinit var channel: BasicMessageChannel<String>
    private var context: Context? = null
    private var orientationEventListener: OrientationEventListener? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = BasicMessageChannel(
            flutterPluginBinding.binaryMessenger,
            "basic_message_channel_example",
            StringCodec.INSTANCE
        )
        setupChannel()
        startOrientationObserver()
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        stopOrientationObserver()
        context = null
    }

    private fun setupChannel() {
        channel.setMessageHandler { message, reply ->
            if (message == "get_orientation") {
                reply.reply(currentOrientation())
            } else {
                reply.reply("Unknown command")
            }
        }
    }

    private fun currentOrientation(): String {
        val orientation = context?.resources?.configuration?.orientation
        return when (orientation) {
            Configuration.ORIENTATION_LANDSCAPE -> "Landscape"
            Configuration.ORIENTATION_PORTRAIT -> "Portrait"
            else -> "Undefined"
        }
    }

    private fun startOrientationObserver() {
        orientationEventListener = object : OrientationEventListener(context) {
            override fun onOrientationChanged(orientation: Int) {
                val info = currentOrientation()
                channel.send("orientation_update:$info")
            }
        }
        orientationEventListener?.enable()
    }

    private fun stopOrientationObserver() {
        orientationEventListener?.disable()
    }
}
